crates/sui-transactional-test-runner/src/test_adapter.rs

```diff
@@ -595,11 +604,151 @@ impl<'a> MoveTestAdapter<'a> for SuiTestAdapter<'a> {
+                                    let (named_addr_opt, _id) = annot_module.module_id();
+                                    let named_addr_opt = named_addr_opt.map(|n| n.value);
+                                    let module = annot_module.named_module.module;
+                                    (named_addr_opt, module)
+                                }
+                                AnnotatedCompiledUnit::Script(_) => panic!(
+                                    "Expected a module text block, not a script, \
+                                    following 'upgrade' starting on lines {}-{}",
+                                    start_line, command_lines_stop
+                                ),
+                            })
+                            .collect();
+                        (modules, warnings_opt)
+                    }
+                    SyntaxChoice::IR => {
+                        let module = compile_ir_module(state, data.path())?;
+                        (vec![(None, module)], None)
+                    }
+                };
+                let (output, mut modules) =
+                    self.upgrade_package(package, modules, upgrade_capability, gas_budget)?;
+                match syntax {
+                    SyntaxChoice::Source => {
+                        let path = data.path().to_str().unwrap().to_owned();
+                        self.compiled_state()
+                            .add_with_source_file(modules, (path, data))
+                    }
+                    SyntaxChoice::IR => {
+                        let module = modules.pop().unwrap().1;
+                        self.compiled_state()
+                            .add_and_generate_interface_file(module);
+                    }
+                };
+                Ok(merge_output(warnings_opt, output))
+            }
         }
     }
 }
 
+fn merge_output(left: Option<String>, right: Option<String>) -> Option<String> {
+    match (left, right) {
+        (None, right) => right,
+        (left, None) => left,
+        (Some(mut left), Some(right)) => {
+            left.push_str(&right);
+            Some(left)
+        }
+    }
+}
+
+macro_rules! move_call {
+    {$builder:expr, ($addr:expr)::$module_name:ident::$func:ident($($args:expr),* $(,)?)} => {
+        $builder.programmable_move_call(
+            $addr,
+            ident_str!(stringify!($module_name)).to_owned(),
+            ident_str!(stringify!($func)).to_owned(),
+            vec![],
+            vec![$($args),*],
+        )
+    }
+}
+
 impl<'a> SuiTestAdapter<'a> {
+    fn upgrade_package(
+        &mut self,
+        current_package_id: Option<String>,
+        modules: Vec<(Option<Symbol>, CompiledModule)>,
+        upgrade_capability: u64,
+        gas_budget: Option<u64>,
+    ) -> anyhow::Result<(Option<String>, Vec<(Option<Symbol>, CompiledModule)>)> {
+        let current_package_id = (*modules[0].1.self_id().address()).into();
```

Comment:
This binding shadows `current_package_id` above, which I think you want to use instead?   The `modules` that you get in at this point should have their self addresses set to `0x0` in preparation for publishing.

Reply:
Another thought, unlike for `publish`, we can't have one optional package name parameter.  We need two:

- One mandatory package name parameter, to identify the package being upgrades.
- One optional parameter to name the upgraded version, because both can co-exist.

Reply:
Gotcha. How do we see the optional parameter referring to other/prior versions? I'm not sure what values to look at to resolve those versions if we had something like `--version V1` versus `--version V2`.

So, would it work to just refer to the `FakeID` of created/upgraded objects like `--previous-package 1,2`? In which case, I'm not sure we need a package name parameter in those cases?

Reply:
also, maybe better for a follow up to this minimally viable setup in this PR?

Reply:
You would refer to the previous version by its package name and optionally give the new package version a name as well, so no need to introduce a new version parameter, it would all be handled using the existing named address system.

Unfortunately, I don't think the current approach is minimally viable -- it may leave the compiled state in an inconsistent state if the upgrade fails and doesn't really permit the kind of calls into both the previous and new versions of a package that are needed to support our tests.  I don't think the fixes there would require a huge addition to this PR though.